### ISSUE TITLE
database not closed if sqlite3_open_v2 fails

### ISSUE THREAD

**Issue by rittneje** [mattn/go-sqlite3#668]
According to the [SQLite documentation](https://www.sqlite.org/c3ref/open.html), the database handle assigned by `sqlite3_open_v2` can be valid even if it returns an error code, and thus you always need to call `sqlite3_close` on it. However, right now the library just returns in this situation, effectively leaking the handle.

**Comment by mattn** [comment]
Probably duplicate #667

**Comment by rittneje** [comment]
@mattn How is this a duplicate of #667?

**Comment by gonapps** [comment]
I've investigated on this problem and found [this](https://stackoverflow.com/questions/32345124/why-does-sql-open-return-nil-as-error-when-it-should-not)
According to this link, any code that returns error in [SQLiteDriver.Open](https://github.com/mattn/go-sqlite3/blob/master/sqlite3.go#L979) doesn't immediately returns error from sql.Open()
To get an error without querying, you should use db.Open()

**Comment by rittneje** [comment]
That's not the bug I'm reporting. You just have to call `sqlite3_close` [here](https://github.com/mattn/go-sqlite3/blob/master/sqlite3.go#L1345), as the documentation prescribes.

**Comment by mattn** [comment]
go-sqlite3 have a flag SQLITE_OPEN_CREATE for sqlite3_open_v2 currently. So always create db file.

**Comment by rittneje** [comment]
@mattn I don't understand what that has to do with this. This bug is not related to #667 in any way.

The SQLite documentation for `sqlite3_open_v2` states: 

> A database connection handle is usually returned in *ppDb, even if an error occurs. [...] **Whether or not an error occurs when it is opened**, resources associated with the database connection handle should be released by passing it to sqlite3_close() when it is no longer required.

However, right now, this library just does:

```go
rv := C._sqlite3_open_v2(name, &db,
    mutex|C.SQLITE_OPEN_READWRITE|C.SQLITE_OPEN_CREATE,
    nil)
if rv != 0 {
    return nil, Error{Code: ErrNo(rv)}
}
```

Consequently, you aren't following the spec. You must call `sqlite3_close` (or `sqlite3_close_v2`) in the case where `rv != 0` or you might leak resources.

**Comment by mattn** [comment]
addressed https://github.com/mattn/go-sqlite3/pull/670

**Comment by mattn** [comment]
closed by #670
